### ISSUE TITLE
Autoloading and Zeitwerk unexpected behaivor with namespaces

### ISSUE THREAD

**Issue by FelipeRiveraC** [rails/rails#50207]
### Steps to reproduce

```ruby
# --- app/models/namespace_a.rb
module NamespaceA
  def self.table_name_prefix
    'namespace_a_'
  end
end

# --- app/models/namespace_a/namespace_b.rb
module NamespaceA
  module NamespaceB
    def self.table_name_prefix
      'namespace_a_namespace_b_'
    end
  end
end

# --- app/models/namespace_a/namespace_b/class_a.rb

class NamespaceA::NamespaceB::ClassA < ApplicationRecord
end

```

### Expected behavior
The expected behaivor is that Zeitwerk load ```ClassA``` as ```NamespaceA::NamespaceB::ClassA```. 

### Actual behavior
(this on rails server, on rails console it works just fine)
For some reason, when Zeitwerk loads ```app/models/namespace_a/namespace_b.rb``` it stops, and dosen´t load ```app/models/namespace_a/namespace_b/class_a.rb```.

If you delete the file ```app/models/namespace_a/namespace_b.rb```, then Zeitwerk will load the ```ClassA``` as ```NamespaceA::NamespaceB::ClassA```.

This error is raised
```
*** NameError Exception: uninitialized constant NamespaceA::NamespaceB

NamespaceA::NamespaceB::ClassA
          ^^^^^^^^^^^^
```
nil

This issue was already discussed [here](https://github.com/rails/rails/issues/38726).

@fxn 

### System configuration
**Rails version**: 7.0.6

**Ruby version**: 3.1

**Comment by fxn** [comment]
I don't understand what you describe.

> when Zeitwerk loads app/models/namespace_a/namespace_b.rb it stops, and dosen´t load app/models/namespace_a/namespace_b/class_a.rb

Why do you believe loading that file should load the other file?

Can you run `bin/rails zeitwerk:check`? Does it give you any error?

**Comment by FelipeRiveraC** [comment]
> I don't understand what you describe.
> 
> > when Zeitwerk loads app/models/namespace_a/namespace_b.rb it stops, and dosen´t load app/models/namespace_a/namespace_b/class_a.rb
> 
> Why do you believe loading the first file should load the second file?
> 
> Can you run `bin/rails zeitwerk:check`? Does it give you any error?

I expect to load the class in that directory and namespace like it apears in rails console (wich suggests me its a autoloading probem)


```
Zeitwerk@rails.main: constant NamespaceA::NamespaceB loaded from file /Users/feliperivera//app/models/namespace_a/namespace_b.rb
Zeitwerk@rails.main: constant NamespaceA::NamespaceB::ClassA loaded from file /Users/feliperivera//app/models/namespace_a/namespace_b/class_a.rb
=> NamespaceA::NamespaceB::ClassA (call 'NamespaceA::NamespaceB::ClassA.connection' to establish a connection)
```

But inside the rails server, using a binding.pry i can verify that it dosent load in that way, it raises the above error.

```bin/rails zeitwerk:check``` gives me ```All is good!```

**Comment by fxn** [comment]
> I expect to load the class in that directory and namespace like it apears in rails console

To be clear, loading the file `pp/models/namespace_a/namespace_b.rb` does not load the child files (unless classes in subdirectories are referenced in the body of the namespace itself). Classes are loaded when they are referenced, not when their namespaces are referenced.

It could be the case that `binding.pry` is conflicting. If `zeitwerk:check` says all is good, all is good re autoloading. The class should work when using the server as it works when using the console.

For context, some libs hooking into processes like debuggers do work, and some don't ([docs](pp/models/namespace_a/namespace_b.rb)).

Could you please provide a minimal application, with the `binding.pry` call in place so that I can download and execute and see the class not being loaded?

**Comment by FelipeRiveraC** [comment]
You are right. When changing the `binding.pry` for a `puts` the class apears to be loaded correctly. This is probably what you said, a conflict with `binding.pry`. 

Thanks for your time.
